### PR TITLE
Frontdoor rules-engine and routing-rule bug fixes

### PR DESCRIPTION
**Pull request by aramesh-msft**
* Cache duration and query parameters options enabled
* rules-engine conditions can be empty, match values can be empty for 'Any' operator
* routing-rule update can enable Caching
* updated help text

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S170

**Comment by aramesh-msft** [comment]
> add to S170

@MyronFanQiu Could you please review this PR when you get a chance? Thanks.

File: src/front-door/azext_front_door/_params.py [hunk 1]
```diff
@@ -276,7 +278,7 @@ def load_arguments(self, _):
         c.argument('front_door_name', frontdoor_name_type, id_part=None)
         c.argument('rules_engine_name', rules_engine_name_type, id_part=None)
         c.argument('rule_name', options_list=['--name', '-n'], help='Name of the rule')
-        c.argument('action_type', arg_group="Action", arg_type=get_enum_type(['RequestHeader', 'ResponseHeader', 'ForwardRouteOverride', 'RedirectRouteOverride']), help='Action type to apply for a rule.')
+        c.argument('action_type', arg_group="Action", arg_type=get_enum_type(['RequestHeader', 'ResponseHeader']), help='Action type to apply for a rule.')
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/_params.py:281]: Are these two parameters no longer supported?

File: src/front-door/azext_front_door/_params.py [hunk 2]
```diff
@@ -290,6 +292,7 @@ def load_arguments(self, _):
         c.argument('match_processing_behavior', arg_type=get_enum_type(MatchProcessingBehavior), help='Whether to stop processing rules after conditions in a rule is satisfied.')
 
     with self.argument_context('network front-door rules-engine rule action', arg_group='Forward Route Override') as c:
+        c.argument('action_type', arg_group="Action", arg_type=get_enum_type(['RequestHeader', 'ResponseHeader', 'ForwardRouteOverride', 'RedirectRouteOverride']), help='Action type to apply for a rule.')
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/_params.py:295]: we still support these two. Which one is correct?

File: src/front-door/azext_front_door/_params.py (hunk 1 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py:281]: It is still supported but only in 'rules-engine rule action add' command below.

File: src/front-door/azext_front_door/_params.py (hunk 2 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py:295]: It is still supported but in 'rules-engine rule create' command only header actions are supported as a starter rule.

**Comment by aramesh-msft** [comment]
I see that this review has one approval but merging is still blocked, @jsntcy could you please take a look and approve/merge this pull request?

File: src/front-door/azext_front_door/custom.py [hunk 3]
```diff
@@ -1131,12 +1145,14 @@ def create_rules_engine_rule(cmd, resource_group_name, front_door_name, rules_en
...
     match_conditions = []
+    if match_values is None:
+        match_values = []
     condition = RulesEngineMatchCondition(rules_engine_match_variable=match_variable,
                                           rules_engine_operator=operator,
                                           rules_engine_match_value=match_values,
                                           selector=selector,
                                           negate_condition=negate_condition, transforms=transforms)
-    if condition is not None:
+    if condition is not None and match_variable is not None:
```
Comment (jsntcy) [inline src/front-door/azext_front_door/custom.py:1155]: >match_variable [](start = 33, length = 14)

Why did you add `and match_variable is not None` here, but didn't add such check in `add_rules_engine_condition` as it seems they have similar logic?

File: src/front-door/azext_front_door/_params.py [hunk 4]
```diff
@@ -129,15 +129,17 @@ def load_arguments(self, _):
...
         c.argument('patterns_to_match', options_list='--patterns', nargs='+', help='Space-separated list of patterns to match. Default: \'/*\'.')
-        c.argument('rules_engine', help='Name or ID of a Rules Engine configuration.', validator=validate_rules_engine)
+        c.argument('rules_engine', help='Name or ID of a Rules Engine configuration. To unlink property, \"--remove rulesEngine\"', validator=validate_rules_engine)
         c.argument('frontend_endpoints', help='Space-separated list of frontend endpoint names or IDs.', nargs='+', validator=validate_frontend_endpoints)
     with self.argument_context('network front-door routing-rule', arg_group='Forward Routing Rule') as c:
         c.argument('backend_pool', help="Name or ID of a backend pool. It's required to create a Forward routing rule.", validator=validate_backend_pool)
-        c.argument('forwarding_protocol', help='Protocol to use for forwarding traffic.')
+        c.argument('forwarding_protocol', arg_type=get_enum_type(FrontDoorForwardingProtocol), help='Protocol to use for forwarding traffic.')
```
Comment (jsntcy) [inline src/front-door/azext_front_door/_params.py:136]: Extract this argument to a CLIArgumentType for reuse as there are multiple places use this argument.

File: src/front-door/azext_front_door/_params.py [hunk 5]
```diff
@@ -129,15 +129,17 @@ def load_arguments(self, _):
...
         c.argument('custom_forwarding_path', help='Custom path used to rewrite resource paths matched by this rule. Leave empty to use incoming path.')
         c.argument('caching', arg_type=get_three_state_flag(positive_label='Enabled', negative_label='Disabled', return_label=False), help='Whether to enable caching for this route.')
+        c.argument('cache_duration', help='The duration for which the content needs to be cached. Allowed format is ISO 8601 duration')
```
Comment (jsntcy) [inline src/front-door/azext_front_door/_params.py:139]: what's the unit for duration? seconds? minutes? better to add it in help so users can understand it directly.

File: src/front-door/azext_front_door/_params.py [hunk 6]
```diff
@@ -129,15 +129,17 @@ def load_arguments(self, _):
     with self.argument_context('network front-door routing-rule', arg_group=None) as c:
         c.argument('accepted_protocols', nargs='+', help='Space-separated list of protocols to accept. Default: Http')
         c.argument('patterns_to_match', options_list='--patterns', nargs='+', help='Space-separated list of patterns to match. Default: \'/*\'.')
-        c.argument('rules_engine', help='Name or ID of a Rules Engine configuration.', validator=validate_rules_engine)
+        c.argument('rules_engine', help='Name or ID of a Rules Engine configuration. To unlink property, \"--remove rulesEngine\"', validator=validate_rules_engine)
```
Comment (jsntcy) [inline src/front-door/azext_front_door/_params.py:132]: Is there an argument "--remove" for the commands? if not, please don't use "--".

File: src/front-door/azext_front_door/custom.py (hunk 3 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/custom.py:1155]: Because "rule create" can be done without any conditions i.e. it acts like a global action that executes all actions without any match condition, this change is to allow that. Whereas the other command to `add_rules_engine_condition` is to specifically add conditions hence this check was not necessary

File: src/front-door/azext_front_door/_params.py (hunk 5 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py:139]: I have mentioned that the duration is ISO 8601 (self explanatory) as it is widely used, it's like P1DT12H30M30S means 1 day, 12 hours, 30 minutes, 30 seconds.

File: src/front-door/azext_front_door/_params.py (hunk 4 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py:136]: I will consider this in the next commit, for now let's get this merged as it is just a make it "pretty"change.

File: src/front-door/azext_front_door/_params.py (hunk 6 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py:132]: There is --remove for the command hence to remove property I have mentioned the same in help as it is used to link or unlink rulesEngine resource.
